### PR TITLE
UCP/CORE: Fix support for Event FD/CB

### PR DESCRIPTION
**Pull request by dmitrygx**
## What

Fix support for Event FD/CB

## Why ?

UCP doesn't arm for events if UCT support async callback invocation event mechanism

## How ?

1. Add all UCTs with FD/CB event mechanisms to the ARM list
2. If user support both Event mechanisms, use CB

### CODE REVIEW

File: src/ucp/core/ucp_worker.c [hunk 1]
```diff
@@ -328,10 +328,26 @@ static void ucp_worker_wakeup_cleanup(ucp_worker_h worker)
     }
 }
 
+static UCS_F_ALWAYS_INLINE
+int ucp_worker_iface_has_event_notify(const ucp_worker_iface_t *wiface)
+{
+    return (wiface->attr.cap.event_flags & UCT_IFACE_FLAG_EVENT_FD) ||
+           (wiface->attr.cap.event_flags & UCT_IFACE_FLAG_EVENT_ASYNC_CB);
```
Comment (yosefe) [inline src/ucp/core/ucp_worker.c:335]: (wiface->attr.cap.event_flags & (UCT_IFACE_FLAG_EVENT_FD|UCT_IFACE_FLAG_EVENT_ASYNC_CB)

File: src/ucp/core/ucp_worker.c [hunk 2]
```diff
@@ -328,10 +328,26 @@ static void ucp_worker_wakeup_cleanup(ucp_worker_h worker)
...
+}
+
+static UCS_F_ALWAYS_INLINE
+int ucp_worker_iface_use_event_fd(const ucp_worker_iface_t *wiface)
+{
+    /* use iface's fd if it is supported by UCT iface and asynchronous
+     * callback mechanism isn't supported (this is preferred mechanism) */
```
Comment (yosefe) [inline src/ucp/core/ucp_worker.c:342]: document that we prefer the async callback since it will be called anyway

File: src/ucp/core/ucp_worker.c [hunk 3]
```diff
@@ -1266,7 +1288,7 @@ void ucp_worker_iface_cleanup(ucp_worker_iface_t *wiface)
     ucp_worker_iface_disarm(wiface);
 
     if ((wiface->event_fd != -1) &&
-        (wiface->attr.cap.event_flags & UCT_IFACE_FLAG_EVENT_FD)) {
+        ucp_worker_iface_use_event_fd(wiface)) {
```
Comment (yosefe) [inline src/ucp/core/ucp_worker.c:1291]: maybe replace the check ucp_worker_iface_use_event_fd(wiface) by ucs_assert()?

File: src/ucp/core/ucp_worker.c [hunk 4]
```diff
@@ -2143,6 +2165,12 @@ ucs_status_t ucp_worker_wait(ucp_worker_h worker)
         pfd = ucs_alloca(sizeof(*pfd) * worker->context->num_tls);
         nfds = 0;
         ucs_list_for_each(wiface, &worker->arm_ifaces, arm_list) {
+            if (wiface->attr.cap.event_flags & UCT_IFACE_FLAG_EVENT_ASYNC_CB) {
```
Comment (yosefe) [inline src/ucp/core/ucp_worker.c:2168]: if (!ucp_worker_iface_use_event_fd() )
 continue

File: src/ucp/core/ucp_worker.c (hunk 1 above)
Comment (dmitrygx) [reply src/ucp/core/ucp_worker.c:335]: done

File: src/ucp/core/ucp_worker.c (hunk 2 above)
Comment (dmitrygx) [reply src/ucp/core/ucp_worker.c:342]: done

File: src/ucp/core/ucp_worker.c (hunk 4 above)
Comment (dmitrygx) [reply src/ucp/core/ucp_worker.c:2168]: done

File: src/ucp/core/ucp_worker.c (hunk 3 above)
Comment (dmitrygx) [reply src/ucp/core/ucp_worker.c:1291]: done

**Comment by dmitrygx** [comment]
@yosefe could you review pls?

**Pull request merged**